src/mimeparser.rs

```diff
@@ -1303,13 +1304,26 @@ fn get_attachment_filename(mail: &mailparse::ParsedMail) -> Result<Option<String
                 }
             });
         if apostrophe_encoded {
-            // we're currently always assuming utf-8, this might need adaption, however, should not break things.
             if let Some(name) = desired_filename {
-                desired_filename = if let Some(name) = name.splitn(3, '\'').last() {
-                    Some(percent_decode_str(&name).decode_utf8_lossy().to_string())
-                } else {
-                    None
-                }
+                let mut parts = name.splitn(3, '\'');
+                desired_filename =
+                    if let (Some(charset), Some(value)) = (parts.next(), parts.last()) {
+                        let decoded_bytes = percent_decode_str(&value);
+                        if charset.to_lowercase() == "utf-8" {
+                            Some(decoded_bytes.decode_utf8_lossy().to_string())
+                        } else {
+                            let charset = match Charset::for_label(charset.as_bytes()) {
+                                Some(c) => c,
+                                // encoded_words crate say, latin-1 is not reported; moreover, latin1 is a good default
+                                None => Charset::for_label(b"latin1").unwrap(),
+                            };
+                            let decoded_bytes = decoded_bytes.collect::<Vec<u8>>();
```

Comment:
Could use something like
```
let decoded_bytes = Cow<[u8]>::from(decoded_bytes);
```
to avoid vector allocation if filename is not percent-encoded.

Reply:
i tried that, played around a bit, i did not manage to get sth. like that compiling.